### PR TITLE
bpo-1635741: Port _dbm module to multiphase initialization

### PR DESCRIPTION
**Pull request by corona10**
https://bugs.python.org/issue1635741

### CODE REVIEW

**Comment by corona10** [comment]
@shihai1991 PTAL!

**Comment by corona10** [comment]
I' ve checked the memory leak with these scripts and no leaks found

```
./python.exe -m test test_dbm  -R 3:3
```

```  python
def test_dbm(self):
        code = textwrap.dedent(r"""
            import glob

            import test.support
            dbm = test.support.import_module('_dbm')
            _fname = test.support.TESTFN

            def delete_files():
                for f in glob.glob(_fname + "*"):
                    test.support.unlink(f)

            f = dbm.open(_fname, 'n')
            f[b'g'] = b"indented"
            f.close()
            delete_files()
        """)
        ret = test.support.run_in_subinterp(code)
        self.assertEqual(ret, 0)
```

File: Modules/_dbmmodule.c [hunk 1]
```diff
@@ -480,42 +491,68 @@ static PyMethodDef dbmmodule_methods[] = {
     { 0, 0 },
 };
 
+static int
+_dbm_exec(PyObject *module)
+{
+    _dbm_state *state = get_dbm_state(module);
+    state->dbm_type = (PyTypeObject *)PyType_FromModuleAndSpec(module, &dbmtype_spec, NULL);
```
Comment (shihai1991) [inline Modules/_dbmmodule.c]: pep7: No line should be longer than 79 characters.

**Review by corona10 — COMMENTED** [review]
@shihai1991 Thanks shihai! I've updated it

**Review by shihai1991 — APPROVED** [review]
Thanks, LGTM.

File: Modules/_dbmmodule.c [hunk 2]
```diff
@@ -78,14 +86,22 @@ dbm_dealloc(dbmobject *dp)
 {
     if ( dp->di_dbm )
         dbm_close(dp->di_dbm);
-    PyObject_Del(dp);
+    PyTypeObject *tp = Py_TYPE(dp);
+    freefunc free_func = PyType_GetSlot(tp, Py_tp_free);
+    free_func(dp);
+    Py_DECREF(tp);
```
Comment (vstinner) [inline Modules/_dbmmodule.c:94]: This change is unrelated to "Port _dbm module to multiphase initialization". I would prefer to see it in a separated PR. Also, currently `type->tp_free` is still preferred. There is no plan to move stdlib C extensions to the limited C API. If you have such plan, please open an issue to discuss this idea.

File: Modules/_dbmmodule.c [hunk 3]
```diff
@@ -373,46 +399,31 @@ dbm__exit__(PyObject *self, PyObject *args)
...
+    {Py_mp_length, dbm_length},
+    {Py_mp_subscript, dbm_subscript},
+    {Py_mp_ass_subscript, dbm_ass_sub},
+    {0, 0}
+};
+
+static PyType_Spec dbmtype_spec = {
+    .name = "_dbm.dbm",
+    .basicsize = sizeof(dbmobject),
+    .flags = Py_TPFLAGS_DEFAULT,
```
Comment (vstinner) [inline Modules/_dbmmodule.c:417]: Can you please add a comment explaining that calling PyType_GetModuleState() is safe, because `Py_TPFLAGS_BASETYPE` flag is not used?

File: Modules/_dbmmodule.c [hunk 4]
```diff
@@ -78,14 +86,22 @@ dbm_dealloc(dbmobject *dp)
...
+    freefunc free_func = PyType_GetSlot(tp, Py_tp_free);
+    free_func(dp);
+    Py_DECREF(tp);
 }
 
 static Py_ssize_t
 dbm_length(dbmobject *dp)
 {
+    PyTypeObject *tp = Py_TYPE(dp);
+    _dbm_state *state =  PyType_GetModuleState(tp);
```
Comment (vstinner) [inline Modules/_dbmmodule.c]: I would prefer `assert(state != NULL);` since this call **cannot** fail with the current code (see my comment about Py_TPFLAGS_BASETYPE).

For example (master branch, without this change):

```
Python 3.10.0a0 (heads/master:25f38d7044, Jun 15 2020, 14:16:18) 
>>> import _dbm
>>> d=_dbm.open('test.db', "c")
>>> dbm_type = type(d)

# legit
>>> dbm_type.setdefault(d, "key", "value")
'value'

# invalid: error
>>> dbm_type.setdefault("string", "key", "value")
TypeError: descriptor 'setdefault' for '_dbm.dbm' objects doesn't apply to a 'str' object
```

It seems like the descriptor checks the "self" type.

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Modules/_dbmmodule.c (hunk 2 above)
Comment (corona10) [reply Modules/_dbmmodule.c:94]: @vstinner Hmm without this change this PR will cause the leak, but is it okay?

**Review by corona10 — COMMENTED** [review]
I have made the requested changes; please review again

File: Modules/_dbmmodule.c (hunk 2 above)
Comment (vstinner) [reply Modules/_dbmmodule.c:94]: You're right, ignore my comment.

I didn't notice that your change also converts a static type into a heap type. You should elaborate your commit messages: list all major changes. This PR is not just "Port _dbm module to multiphase initialization".

File: Modules/_dbmmodule.c [hunk 5]
```diff
@@ -76,16 +84,25 @@ newdbmobject(const char *file, int flags, int mode)
...
+    assert(state != NULL);
+    if (state == NULL) {
+        return -1;
+    }
```
Comment (vstinner) [inline Modules/_dbmmodule.c]: Well, I propose to replace these checks with `assert(state != NULL);` for all PyType_GetModuleState() calls, not only in this function.

File: Modules/_dbmmodule.c [hunk 6]
```diff
@@ -373,46 +400,36 @@ dbm__exit__(PyObject *self, PyObject *args)
...
+};
+
+
+static PyType_Spec dbmtype_spec = {
+    .name = "_dbm.dbm",
+    .basicsize = sizeof(dbmobject),
+    /*
+    Calling PyType_GetModuleState() is safe
+    because Py_TPFLAGS_BASETYPE flag is not used.
+    */
```
Comment (vstinner) [inline Modules/_dbmmodule.c]: I suggest to elaborate a bit:

```suggestion
    // Calling PyType_GetModuleState() on a subclass is not safe.
    // dbmtype_spec does not have Py_TPFLAGS_BASETYPE flag which prevents to create a subclass.
    // So calling PyType_GetModuleState() in this file is always safe.
```

(I let you reformat the comment ;-))

File: Modules/_dbmmodule.c [hunk 7]
```diff
@@ -76,16 +84,22 @@ newdbmobject(const char *file, int flags, int mode)
...
+    PyTypeObject *tp = Py_TYPE(dp);
+    tp->tp_free(dp);
+    Py_DECREF(tp);
 }
 
 static Py_ssize_t
 dbm_length(dbmobject *dp)
 {
+    PyTypeObject *tp = Py_TYPE(dp);
+    _dbm_state *state =  PyType_GetModuleState(tp);
```
Comment (vstinner) [inline Modules/_dbmmodule.c]: I'm not sure that tp variable is worth it:

```suggestion
    _dbm_state *state = PyType_GetModuleState(Py_TYPE(dp));
```

nitpick: my proposed change also removed an useless whitespace :-D

File: Modules/_dbmmodule.c (hunk 6 above)
Comment (corona10) [reply Modules/_dbmmodule.c]: Applied!

File: Modules/_dbmmodule.c [hunk 8]
```diff
@@ -43,28 +56,25 @@ typedef struct {
...
-static PyTypeObject Dbmtype;
-
-#define is_dbmobject(v) Py_IS_TYPE(v, &Dbmtype)
-#define check_dbmobject_open(v) if ((v)->di_dbm == NULL) \
-               { PyErr_SetString(DbmError, "DBM object has already been closed"); \
-                 return NULL; }
-
-static PyObject *DbmError;
+#define check_dbmobject_open(v, err)                            \
+if ((v)->di_dbm == NULL) {                                      \
```
Comment (vstinner) [inline Modules/_dbmmodule.c]: nitpick: add one indentation level for the if { ... } block :-)

File: Modules/_dbmmodule.c [hunk 9]
```diff
@@ -480,42 +489,69 @@ static PyMethodDef dbmmodule_methods[] = {
...
+                                                        &dbmtype_spec, NULL);
+    if (state->dbm_type == NULL) {
+        return -1;
+    }
+    state->dbm_error = PyErr_NewException("_dbm.error", PyExc_OSError, NULL);
+    if (PyModule_AddStringConstant(module, "library", which_dbm) < 0) {
+        return -1;
+    }
+    Py_INCREF(state->dbm_error);
+    if (PyModule_AddObject(module, "error", state->dbm_error) < 0) {
```
Comment (vstinner) [inline Modules/_dbmmodule.c]: PyModule_AddType() can be used here :-)

File: Modules/_dbmmodule.c [hunk 10]
```diff
@@ -480,42 +489,69 @@ static PyMethodDef dbmmodule_methods[] = {
...
+static int
+_dbm_exec(PyObject *module)
+{
+    _dbm_state *state = get_dbm_state(module);
+    state->dbm_type = (PyTypeObject *)PyType_FromModuleAndSpec(module,
+                                                        &dbmtype_spec, NULL);
+    if (state->dbm_type == NULL) {
+        return -1;
+    }
+    state->dbm_error = PyErr_NewException("_dbm.error", PyExc_OSError, NULL);
```
Comment (vstinner) [inline Modules/_dbmmodule.c:502]: Hum, PyErr_NewException() can fail. Would you mind to check for error?

File: Modules/_dbmmodule.c [hunk 11]
```diff
@@ -373,46 +386,35 @@ dbm__exit__(PyObject *self, PyObject *args)
...
-    0,                        /*tp_iter*/
-    0,                        /*tp_iternext*/
-    dbm_methods,          /*tp_methods*/
+
+static PyType_Spec dbmtype_spec = {
+    .name = "_dbm.dbm",
+    .basicsize = sizeof(dbmobject),
+    // Calling PyType_GetModuleState() on a subclass is not safe.
+    // dbmtype_spec does not have Py_TPFLAGS_BASETYPE flag which prevents to create a subclass.
+    // So calling PyType_GetModuleState() in this file is always safe.
```
Comment (vstinner) [inline Modules/_dbmmodule.c:416]: nitpick: Maybe start to reformat to fit into 80 columns.

File: Modules/_dbmmodule.c [hunk 12]
```diff
@@ -443,19 +445,26 @@ dbmopen_impl(PyObject *module, PyObject *filename, const char *flags,
...
+    if (strcmp(flags, "r") == 0) {
         iflags = O_RDONLY;
-    else if ( strcmp(flags, "w") == 0 )
+    }
+    else if (strcmp(flags, "w") == 0) {
         iflags = O_RDWR;
-    else if ( strcmp(flags, "rw") == 0 ) /* B/W compat */
+    }
+    else if (strcmp(flags, "rw") == 0) {
+        /* B/W compat */
```
Comment (vstinner) [inline Modules/_dbmmodule.c]:
```suggestion
        /* Backward compatibility */
```

**Review by vstinner — COMMENTED** [review]
Thanks! It's now much better!

**Review by vstinner — APPROVED** [review]
LGTM, thanks!
